docs/config/01-configuration-file.md
**In order to serve you well, Karma needs to know about your project in order to test it
and this is done via a configuration file. This page explains how to create such a configuration file.**


Note: Most of the framework adapters, reporters, preprocessors and launchers needs to be loaded as [plugins].


The Karma configuration file can be written in JavaScript or CoffeeScript and is loaded as a regular Node.js module.
Within the configuration file, the configuration code is put together by setting `module.exports` to point to a function
which accepts one argument: the configuration object.

```javascript
// an example karma.conf.js
module.exports = function(config) {
  config.set({
    basePath: '../..',
    frameworks: ['jasmine'],
    //...
  });
};
```

```coffeescript
# an example karma.conf.coffee
module.exports = (config) ->
  config.set
    basePath: '../..'
    frameworks: ['jasmine']
    # ...
```

To see a more detailed example configuration file, see [test/client/karma.conf.js] which contains most of
the available configuration options.

The rest of the document contains a list of all available configuration options...

## autoWatch
**Type:** Boolean

**Default:**  `false`

**CLI:** `--auto-watch`, `--no-auto-watch`

**Description:** Enable or disable watching files and executing the tests
  whenever one of these files changes.


## autoWatchBatchDelay
**Type:** Number

**Default:**  `250`

**Description:** When Karma is watching the files for changes, it tries to batch
multiple changes into a single run so that the test runner doesn't try to start and restart running
tests more than it should. The configuration setting tells Karma how long to wait (in milliseconds) after any changes
have occurred before starting the test process again.


## basePath
**Type:** String

**Default:** `''`

**Description:** The root path location that will be used to resolve all relative
paths defined in `files` and `exclude`. If the `basePath` configuration is a
relative path then it will be resolved to the `__dirname` of the configuration file.


## browsers
**Type:** Array

**Default:**  `[]`

**CLI:** `--browsers Chrome,Firefox`

**Possible Values:**

  * `Chrome` (comes installed with Karma)
  * `ChromeCanary` (comes installed with Karma)
  * `PhantomJS` (comes installed with Karma)
  * `Firefox` (requires karma-firefox-launcher plugin)
  * `Opera` (requires karma-opera-launcher plugin)
  * `Safari` (requires karma-ie-launcher plugin)

**Description:** A list of browsers to launch and capture. When Karma starts up, it will also start up each browser
which is placed within this setting.  Once Karma is shut down, it will shut down these
browsers as well. You can capture any browser manually just by opening the browser and visiting the URL where
the Karma web server is listening (by default it is `http://localhost:9876/`).

See [config/browsers] for more. Additional launchers can be defined through [plugins].


## captureTimeout
**Type:** Number

**Default:** `60000`

**Description:** Timeout for capturing a browser (in ms).

The `captureTimeout` value represents the maximum boot-up time allowed for a browser to start and connect to Karma.
If any browser does not get captured within the timeout, Karma will kill it and try to launch
it again and, after three attempts to capture it, Karma will give up.


## colors
**Type:** Boolean

**Default:**   `true`

**CLI:** `--colors`, `--no-colors`

**Description:**  Enable or disable colors in the output (reporters and logs).


## exclude
**Type:** Array

**Default:** `[]`

**Description:** List of files/patterns to exclude from loaded files.


## files
**Type:** Array

**Default:** `[]`

**Description:** List of files/patterns to load in the browser.

See [config/files] for more information.


## frameworks
**Type:** Array

**Default:** `[]`

**Description:** List of frameworks you want to use. Typically, you will set this to `['jasmine']`, `['mocha']` or `['qunit']`...

Please note just about all frameworks in Karma require an additional plugin/framework library to be installed (via NPM).

Additional information can be found in [plugins].


## hostname
**Type:** String

**Default:** `'localhost'`

**Description:** Hostname to be used when capturing browsers.


## logLevel
**Type:** Constant

**Default:** `config.LOG_INFO`

**CLI:** `--log-level debug`

**Possible values:**

  * `config.LOG_DISABLE`
  * `config.LOG_ERROR`
  * `config.LOG_WARN`
  * `config.LOG_INFO`
  * `config.LOG_DEBUG`

**Description:** Level of logging.


## loggers
**Type:** Array

**Default:** `[{type: 'console'}]`

**Description:** A list of log appenders to be used. See the
  documentation for [log4js] for more information.


## plugins
**Type:** Array

**Default:** `['karma-*']`

**Description:** List of plugins to load. A plugin can be a string (in which case it will be required by Karma) or an inlined plugin - Object.
By default, Karma loads all siblink modules, that match `karma-*`.

Please note just about all plugins in Karma require an additional library to be installed (via NPM).

See [plugins] for more information.


## port
**Type:** Number

**Default:** `9876`

**CLI:** `--port 9876`

**Description:** The port where the webserver will be listening.


## preprocessors
**Type:** Object

**Default:** `{'**/*.coffee': 'coffee'}`

**Description:** A map of preprocessors to use.

Preprocessors can be loaded through [plugins].

Please note just about all preprocessors in Karma (other than CoffeeScript and some other defaults)
require an additional library to be installed (via NPM).

Click <a href="preprocessors.html">here</a> for more information.


## proxies
**Type:** Object

**Default:**   `{}`

**Description:** A map of path-proxy pairs.

**Example:**
  ```javascript
  proxies:  {
    '/static': 'http://gstatic.com',
    '/web': 'http://localhost:9000'
  },
  ```

## proxyValidateSSL
**Type:** Boolean

**Default:**   `true`

**Description:** Whether or not karma or any browsers should raise an error when an inavlid SSL certificate is found.

## reportSlowerThan
**Type:** Number

**Default:** `0`

**Description:** Karma will report all the tests that are slower than given time limit (in ms).
This is disabled by default (since the default value is 0).


## reporters
**Type:** Array

**Default:** `['progress']`

**CLI:** `--reporters progress,growl`

**Possible Values:**

  * `dots`
  * `progress`

**Description:** A list of reporters to use.

Additional reporters, such as `growl`, `junit`, `teamcity` or `coverage` can be loaded through [plugins].

Please note just about all additional reporters in Karma (other than progress) require an additional library to be installed (via NPM).


## singleRun
**Type:** Boolean

**Default:** `false`

**CLI:** `--single-run`, `no-single-run`

**Description:** Continuous Integration mode.

If `true`, Karma will start and capture all configured browsers, run tests and then exit with an exit code of `0` or `1` depending
on whether all tests passed or any tests failed.


## singleRunBatchSize
**Type:** Number

**Default:** `0`

**CLI:** `--single-run-batch-size`

**Description:** Run tests in batches in Continuous Integration mode.

If greater than zero, Karma will start all configured browsers and run tests
in batches of the given size. Karma will exit with an exit code of `0` or `1` depending on whether all batches passed or any batches failed.


## transports
**Type:** Array

**Default:** `['websocket', 'flashsocket', 'xhr-polling', 'jsonp-polling']`

**Description:** An array of allowed transport methods between the browser and testing server. This configuration setting
is handed off to [socket.io](https://github.com/LearnBoost/Socket.IO/wiki/Configuring-Socket.IO) (which manages the communication
between browsers and the testing server).


## urlRoot
**Type:** String

**Default:** `'/'`

**Description:** The base url, where Karma runs.

All the Karma's urls get prefixed with the `urlRoot`. This is helpful when using proxies, as
sometimes you might want to proxy a url that is already taken by Karma.


[plugins]: plugins.html
[test/client/karma.conf.js]: https://github.com/karma-runner/karma/blob/master/test/client/karma.conf.js
[config/files]: files.html
[config/browsers]: browsers.html
[config/preprocessors]: preprocessors.html
[log4js]: https://github.com/nomiddlename/log4js-node


lib/config.js
var path = require('path');

var logger = require('./logger');
var log = logger.create('config');
var helper = require('./helper');
var constant = require('./constants');

// Coffee is required here to enable config files written in coffee-script.
// It's not directly used in this file.
require('coffee-script');


var Pattern = function(pattern, served, included, watched) {
  this.pattern = pattern;
  this.served = helper.isDefined(served) ? served : true;
  this.included = helper.isDefined(included) ? included : true;
  this.watched = helper.isDefined(watched) ? watched : true;
};

var UrlPattern = function(url) {
  Pattern.call(this, url, false, true, false);
};


var createPatternObject = function(pattern) {
  if (helper.isString(pattern)) {
    return helper.isUrlAbsolute(pattern) ? new UrlPattern(pattern) : new Pattern(pattern);
  }

  if (helper.isObject(pattern)) {
    if (!helper.isDefined(pattern.pattern)) {
      log.warn('Invalid pattern %s!\n\tObject is missing "pattern" property".', pattern);
    }

    return helper.isUrlAbsolute(pattern.pattern) ?
        new UrlPattern(pattern.pattern) :
        new Pattern(pattern.pattern, pattern.served, pattern.included, pattern.watched);
  }

  log.warn('Invalid pattern %s!\n\tExpected string or object with "pattern" property.', pattern);
  return new Pattern(null, false, false, false);
};

var normalizeConfig = function(config, configFilePath) {

  var basePathResolve = function(relativePath) {
    if (helper.isUrlAbsolute(relativePath)) {
      return relativePath;
    }

    if (!helper.isDefined(config.basePath) || !helper.isDefined(relativePath)) {
      return '';
    }
    return path.resolve(config.basePath, relativePath);
  };

  var createPatternMapper = function(resolve) {
    return function(objectPattern) {
      objectPattern.pattern = resolve(objectPattern.pattern);

      return objectPattern;
    };
  };

  if (helper.isString(configFilePath)) {
    // resolve basePath
    config.basePath = path.resolve(path.dirname(configFilePath), config.basePath);

    // always ignore the config file itself
    config.exclude.push(configFilePath);
  } else {
    config.basePath = path.resolve(config.basePath || '.');
  }

  config.files = config.files.map(createPatternObject).map(createPatternMapper(basePathResolve));
  config.exclude = config.exclude.map(basePathResolve);
  config.junitReporter.outputFile = basePathResolve(config.junitReporter.outputFile);
  config.coverageReporter.dir = basePathResolve(config.coverageReporter.dir);

  // normalize paths on windows
  config.basePath = helper.normalizeWinPath(config.basePath);
  config.files = config.files.map(createPatternMapper(helper.normalizeWinPath));
  config.exclude = config.exclude.map(helper.normalizeWinPath);
  config.junitReporter.outputFile = helper.normalizeWinPath(config.junitReporter.outputFile);
  config.coverageReporter.dir = helper.normalizeWinPath(config.coverageReporter.dir);

  // normalize urlRoot
  var urlRoot = config.urlRoot;
  if (urlRoot.charAt(0) !== '/') {
    urlRoot = '/' + urlRoot;
  }

  if (urlRoot.charAt(urlRoot.length - 1) !== '/') {
    urlRoot = urlRoot + '/';
  }

  if (urlRoot !== config.urlRoot) {
    log.warn('urlRoot normalized to "%s"', urlRoot);
    config.urlRoot = urlRoot;
  }

  if (config.proxies && config.proxies.hasOwnProperty(config.urlRoot)) {
    log.warn('"%s" is proxied, you should probably change urlRoot to avoid conflicts',
        config.urlRoot);
  }

  if (config.singleRun && config.autoWatch) {
    log.debug('autoWatch set to false, because of singleRun');
    config.autoWatch = false;
  }

  if (helper.isString(config.reporters)) {
    config.reporters = config.reporters.split(',');
  }

  // TODO(vojta): remove
  if (helper.isDefined(config.reporter)) {
    log.warn('"reporter" is deprecated, use "reporters" instead');
  }

  // normalize preprocessors
  var preprocessors = config.preprocessors || {};
  var normalizedPreprocessors = config.preprocessors = Object.create(null);

  Object.keys(preprocessors).forEach(function(pattern) {
    var normalizedPattern = helper.normalizeWinPath(basePathResolve(pattern));

    normalizedPreprocessors[normalizedPattern] = helper.isString(preprocessors[pattern]) ?
        [preprocessors[pattern]] : preprocessors[pattern];
  });


  // define custom launchers/preprocessors/reporters - create an inlined plugin
  var module = Object.create(null);
  ['launcher', 'preprocessor', 'reporter'].forEach(function(type) {
    var definitions = config['custom' + helper.ucFirst(type) + 's'] || {};

    Object.keys(definitions).forEach(function(name) {
      var definition = definitions[name];

      if (!helper.isObject(definition)) {
        return log.warn('Can not define %s %s. Definition has to be an object.', type, name);
      }

      if (!helper.isString(definition.base)) {
        return log.warn('Can not define %s %s. Missing base %s.', type, name, type);
      }

      var token = type + ':' + definition.base;
      var locals = {
        args: ['value', definition]
      };

      module[type + ':' + name] = ['factory', function(injector) {
        return injector.createChild([locals], [token]).get(token);
      }];
    });
  });

  config.plugins.push(module);

  return config;
};

// TODO(vojta): remove in 0.11
var CONST_ERR = '%s is not supported anymore.\n\tPlease use `frameworks = ["%s"];` instead.';
['JASMINE', 'MOCHA', 'QUNIT'].forEach(function(framework) {
  [framework, framework + '_ADAPTER'].forEach(function(name) {
    Object.defineProperty(global, name, {configurable: true, get: function() {
      log.warn(CONST_ERR, name, framework.toLowerCase());
      return __dirname + '/../../karma-' + framework.toLowerCase() + '/lib/' +
          (framework === name ? framework.toLowerCase() : 'adapter') + '.js';
    }});
  });
});

['REQUIRE', 'REQUIRE_ADAPTER'].forEach(function(name) {
  Object.defineProperty(global, name, {configurable: true, get: function() {
    log.warn(CONST_ERR, name, 'requirejs');
    return __dirname + '/../../karma-requirejs/lib/' +
          (name === 'REQUIRE' ? 'require' : 'adapter') + '.js';
  }});
});

['ANGULAR_SCENARIO', 'ANGULAR_SCENARIO_ADAPTER'].forEach(function(name) {
  Object.defineProperty(global, name, {configurable: true, get: function() {
    log.warn(CONST_ERR, name, 'ng-scenario');
    return __dirname + '/../../karma-ng-scenario/lib/' +
          (name === 'ANGULAR_SCENARIO' ? 'angular-scenario' : 'adapter') + '.js';
  }});
});

['LOG_DISABLE', 'LOG_INFO', 'LOG_DEBUG', 'LOG_WARN', 'LOG_ERROR'].forEach(function(name) {
  Object.defineProperty(global, name, {configurable: true, get: function() {
    log.warn('%s is not supported anymore.\n  Please use `karma.%s` instead.', name, name);
    return constant[name];
  }});
});

var Config = function() {
  var config = this;

  this.LOG_DISABLE = constant.LOG_DISABLE;
  this.LOG_ERROR = constant.LOG_ERROR;
  this.LOG_WARN = constant.LOG_WARN;
  this.LOG_INFO = constant.LOG_INFO;
  this.LOG_DEBUG = constant.LOG_DEBUG;

  this.set = function(newConfig) {
    Object.keys(newConfig).forEach(function(key) {
      config[key] = newConfig[key];
    });
  };

  // TODO(vojta): remove this in 0.10
  this.configure = function(newConfig) {
    log.warn('config.configure() is deprecated, please use config.set() instead.');
    this.set(newConfig);
  };

  // TODO(vojta): remove this in 0.10
  ['launcher', 'reporter', 'preprocessor'].forEach(function(type) {
    var methodName = 'define' + helper.ucFirst(type);
    var propertyName = 'custom' + helper.ucFirst(type) + 's';

    config[methodName] = function(name, base, options) {
      log.warn('config.%s is deprecated, please use "%s" instead.', methodName, propertyName);

      if (!helper.isString(name)) {
        return log.warn('Can not define %s. Name has to be a string.', type);
      }

      if (!helper.isString(base)) {
        return log.warn('Can not define %s %s. Missing parent %s.', type, name, type);
      }

      if (!helper.isObject(options)) {
        return log.warn('Can not define %s %s. Arguments has to be an object.', type, name);
      }

      config[propertyName] = config[propertyName] || {};
      config[propertyName][name] = options;
      options.base = base;
    };
  });


  // DEFAULT CONFIG
  this.frameworks = [];
  this.port = constant.DEFAULT_PORT;
  this.hostname = constant.DEFAULT_HOSTNAME;
  this.basePath = '';
  this.files = [];
  this.exclude = [];
  this.logLevel = constant.LOG_INFO;
  this.colors = true;
  this.autoWatch = false;
  this.autoWatchBatchDelay = 250;
  this.reporters = ['progress'];
  this.singleRun = false;
  this.singleRunBatchSize = 0;
  this.browsers = [];
  this.captureTimeout = 60000;
  this.proxies = {};
  this.proxyValidateSSL = true;
  this.preprocessors = {'**/*.coffee': 'coffee', '**/*.html': 'html2js'};
  this.urlRoot = '/';
  this.reportSlowerThan = 0;
  this.loggers = [constant.CONSOLE_APPENDER];
  this.transports = ['websocket', 'flashsocket', 'xhr-polling', 'jsonp-polling'];
  this.plugins = ['karma-*'];
  this.client = {
    args: []
  };

  // TODO(vojta): remove in 0.10
  this.junitReporter = {
    outputFile: 'test-results.xml',
    suite: ''
  };

  // TODO(vojta): remove in 0.10
  this.coverageReporter = {
    type: 'html',
    dir: 'coverage'
  };
};

var CONFIG_SYNTAX_HELP = '  module.exports = function(config) {\n' +
                         '    config.set({\n' +
                         '      // your config\n' +
                         '    });\n' +
                         '  };\n';

var parseConfig = function(configFilePath, cliOptions) {
  var configModule;
  if (configFilePath) {
    log.debug('Loading config %s', configFilePath);

    try {
      configModule = require(configFilePath);
    } catch(e) {
      if (e.code === 'MODULE_NOT_FOUND' && e.message.indexOf(configFilePath) !== -1) {
        log.error('File %s does not exist!', configFilePath);
      } else {
        log.error('Invalid config file!\n  ' + e.stack);
      }
      return process.exit(1);
    }
    if (!helper.isFunction(configModule)) {
      log.error('Config file must export a function!\n' + CONFIG_SYNTAX_HELP);
      return process.exit(1);
    }
  } else {
    // if no config file path is passed, we define a dummy config module.
    configModule = function() {};
  }

  var config = new Config();

  try {
    configModule(config);
  } catch(e) {
    log.error('Error in config file!\n', e);
    return process.exit(1);
  }

  // merge the config from config file and cliOptions (precendense)
  config.set(cliOptions);

  // configure the logger as soon as we can
  logger.setup(config.logLevel, config.colors, config.loggers);

  return normalizeConfig(config, configFilePath);
};

// PUBLIC API
exports.parseConfig = parseConfig;
exports.Pattern = Pattern;
exports.createPatternObject = createPatternObject;
